float32 value
int32 mode
